File: DFS/572.subtree-of-another-tree.go
package leetcode

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// @leet start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSame(l, r *TreeNode) bool {
	if l == nil && r == nil {
		return true
	}
	if l == nil || r == nil {
		return false
	}
	if l.Val != r.Val {
		return false
	}
	return isSame(l.Left, r.Left) && isSame(l.Right, r.Right)
}

func isSubtree(root *TreeNode, subRoot *TreeNode) bool {
	if root == nil {
		return false
	}
	return isSame(root, subRoot) || isSubtree(root.Left, subRoot) || isSubtree(root.Right, subRoot)
}

// @leet end

File: DFS/572.subtree-of-another-tree.ts
import { TreeNode} from "./type" 

// @leet start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */


function isSubtree(root: TreeNode | null, subRoot: TreeNode | null): boolean {
  function isSame(l:TreeNode | null,r:TreeNode|null ){
    if(l == null && r == null) return true
    if(l == null || r == null) return false
    if(l.val !== r.val) return false
    return isSame(l.left, r.left) && isSame(l.right, r.right)
  }

  if(root == null) return false
  return isSame(root, subRoot) || isSubtree(root.left, subRoot)||isSubtree(root.right, subRoot)
    
};
// @leet end
